flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -0,0 +1,279 @@
+
+			while (!taskManagerServices.getJobManagerTable().contains(jobId)) {
+				Thread.sleep(50);
+			}
+
+			final ExecutionAttemptID eid1 = new ExecutionAttemptID();
+			final IntermediateResultPartitionID partitionId = new IntermediateResultPartitionID();
+			final ResultPartitionID resultPartitionId = new ResultPartitionID(partitionId, eid1);
+			final ResultPartitionDeploymentDescriptor task1ResultPartitionDescriptor =
+				new ResultPartitionDeploymentDescriptor(
+					new PartitionDescriptor(new IntermediateDataSetID(), partitionId, ResultPartitionType.BLOCKING, 1, 0),
+					new ShuffleDescriptor() {
+						@Override
+						public ResultPartitionID getResultPartitionID() {
+							return resultPartitionId;
+						}
+					},
+					1,
+					true);
+
+			final TaskDeploymentDescriptor taskDeploymentDescriptor =
+				TaskExecutorSubmissionTest.createTaskDeploymentDescriptor(
+					jobId,
+					"job",
+					eid1,
+					new SerializedValue<>(new ExecutionConfig()),
+					"Sender",
+					1,
+					0,
+					1,
+					0,
+					new Configuration(),
+					new Configuration(),
+					TestInvokable.class.getName(),
+					Collections.singletonList(task1ResultPartitionDescriptor),
+					Collections.emptyList(),
+					Collections.emptyList(),
+					Collections.emptyList(),
+					0);
+
+			taskSlotTable.allocateSlot(0, jobId, taskDeploymentDescriptor.getAllocationId(), Time.seconds(60));
+
+			TestInvokable.sync = new BlockerSync();
+
+			taskManager.submitTask(taskDeploymentDescriptor, jobMasterGateway.getFencingToken(), timeout)
+				.get();
+
+			TestInvokable.sync.awaitBlocker();
+
+			// the task is still running, so nothing was actually stored yet
+			assertTrue(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+
+			TestInvokable.sync.releaseBlocker();
+			taskFinishedFuture.get(timeout.getSize(), timeout.getUnit());
+
+			// the task is still running, so nothing was actually stored yet
+			assertTrue(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+
+			releaseAction.accept(jobId, new ResultPartitionID(partitionId, eid1), taskManager);
+			postReleaseActionAssertion.accept(jobId, jobAwareShuffleEnvironment);
+		} finally {
+			RpcUtils.terminateRpcEndpoint(taskManager, timeout);
+		}
+
+		// the partition table should always be cleaned up on shutdown
+		assertFalse(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+	}
+
+	/**
+	 * Test invokable which completes the given future when executed.
+	 */
+	public static class TestInvokable extends AbstractInvokable {
+
+		static BlockerSync sync;
+
+		public TestInvokable(Environment environment) {
+			super(environment);
+		}
+
+		@Override
+		public void invoke() throws Exception {
+			sync.block();
+		}
+	}
+
+	private TestingTaskExecutor createTestingTaskExecutor(TaskManagerServices taskManagerServices, HeartbeatServices heartbeatServices) throws IOException {
```

Comment:
We could remove `HeartbeatServices` from parameter because all the usages are creating the same `new HeartbeatServices(Long.MAX_VALUE, Long.MAX_VALUE))`